### ISSUE TITLE
Floating point model runs even faster than Binarized model created by larq.

### ISSUE THREAD

**Issue by HuJiawei24** [larq/larq#496]
I used Larq to create a binary resnet20 and a full precision resnet20 by setting the input quantiser and kernel quantiser to 'set_sign' and None respectively. I then converted the .h5 file into tflite files which are subsequently executed on raspberry pi4. The full precision model is even faster than the binary model, but the tflite file size looks correct as the full precision model file is much bigger than the binary model file. I also noticed that when I created the binary model with larq, the weight associated with the model is floating point instead of binary. Is this the latent weight used for training? it is also the weight when I save the model by model.save().

**Comment by leonoverweel** [comment]
Hi, thanks for checking out Larq! Let's see if we can get this sorted for you.

> I used Larq to create a binary resnet20 and a full precision resnet20 by setting the input quantiser and kernel quantiser to 'set_sign' and None respectively.

If you set `input_quantizer="ste_sign"` and `kernel_quantizer=None`, you actually end up with a Binary Activation Network (BAN), not a Binarized Neural Network (BNN)—your activations are binary, but your weights are full precision. We do not have optimized kernels for BANs, so this may explain the speed difference on your Raspberry Pi.

> I then converted the .h5 file into tflite files which are subsequently executed on raspberry pi4. 

Just to make sure: did you use [larq/compute-engine](https://github.com/larq/compute-engine) to do this conversion? If yes, which version of LCE did you use? (Also, which version of Larq are you on?)

> The full precision model is even faster than the binary model, but the tflite file size looks correct as the full precision model file is much bigger than the binary model file.

The size surprises me a bit. Since you set `kernel_quantizer=None`, your weights should still be full precision, not binary, so your binary (BAN) model file shouldn't be smaller than your full precision model file (assuming that you're comparing your binary (BAN) flatbuffer/tflite file to a full precision flatbuffer, not to the .h5 file).

> I also noticed that when I created the binary model with larq, the weight associated with the model is floating point instead of binary. Is this the latent weight used for training? it is also the weight when I save the model by model.save().

Yes, we store full precision latent weights in the model for training; these will also get saved by `model.save()` to make it easy to save and resume from checkpoints during training. You can also see/save the binary weights using our `quantized_scope()` context. Here's a snippet [from the docs](https://docs.larq.dev/larq/guides/bnn-optimization/#retrieving-the-binary-weights) showing how to do that:

```python
model.save("full_precision_model.h5")  # save full precision latent weights
fp_weights = model.get_weights()  # get latent weights

with larq.context.quantized_scope(True):
    model.save("binary_model.h5")  # save binary weights
    weights = model.get_weights()  # get binary weights
```

**Comment by HuJiawei24** [comment]
Thank you for your speedy reply!!! I am sorry that phrased wrongly in my previous post. When I create BNN, I set both "input_quantizer" and "kernel_quantizer" to "ste_sign". On the other hand, I set "input_quantizer" and "kernel_quantizer" to None when I create full precision model. I did use larq/compute-engine to make the conversion. The converted file sizes look correct to me as the binary file size is much smaller than the full precision file, but the speed of the BNN model is slower than the full precision model. I just installed larq/LCE two days ago, so it should be the latest version?

**Comment by leonoverweel** [comment]
No problem, I see, that makes sense. This is hard to diagnose without seeing the model graphs; could you share the .h5 and .tflite files for the binary model so we can take a closer look? Thanks. And yeah, those versions of Larq/LCE should be recent enough.

**Comment by HuJiawei24** [comment]
Again, thank you for your speedy reply!! The .h5 and tflite file is uploaded in the following google drive. I have to thank you for your kind help. This is quite important and urgent to me...

https://drive.google.com/drive/folders/1WeC3t-8wuqByItaqoMgKrmGogDFAlecT

**Comment by HuJiawei24** [comment]
The previous google drive contains the binary model created using larq. I also upload the corresponding full precision model in this link: 
https://drive.google.com/drive/folders/10TV6pBGt7hGM_WJFYL1AUM5tLKjAR_rv

**Comment by leonoverweel** [comment]
Hey, I took a look at your model (`binary_resnet20.h5`) and here are some of my notes:

* Have you trained this network? I don't think it'd be able to get a very high accuracy in its current form, because of my next few notes.
* In general, the first layer of BNNs should be full-precision; see all models in [`larq/zoo`](https://github.com/larq) and the literature behind them, for example Sec. 2.1 of [Courbariaux et al. (2016)](https://arxiv.org/pdf/1602.02830.pdf) and Sec. 3.2 of [Rastegari et al. (2016)](https://arxiv.org/pdf/1603.05279.pdf).
* Having Batch Normalization between binary layers is quite important for making BNNs work; some of the speed in our kernels also comes from the fact that we fuse these into the binary convolutions.
* We have found that binary skip connections are not very useful for model performance; binary skip connections with binary convolutions, like `quant_conv2d_5232` in your graph for example, also are not something we've had success with in our experimentation.

Before looking at inference speed, I think it's better to look at these architecture issues first; solving those has a good chance of improving inference speed too.

Is there any specific reason you want to use this specific resnet20 architecture? If not, I would recommend you take a look at our optimized [`QuickNet`](https://docs.larq.dev/zoo/api/sota/#quicknet) family of architectures and either use one of those, or use it as starting point for designing your own block and network architecture. See [here](https://lutzroeder.github.io/netron/?url=https://cors-anywhere.herokuapp.com/https://github.com/larq/zoo/releases/download/quicknet-v0.2.0/quicknet.json) for a model graph of QuickNet; there are more linked [in the docs](https://docs.larq.dev/zoo/api/sota).

**Comment by HuJiawei24** [comment]
Thank you for your time to help me to analyse the model. I didn’t train the network because I just want to see the speed/inference difference between the binary model and full precision model. It is also the reason why I didn't add in batch norm layer in my network. Regarding your third point, if I could understand you correctly, you're saying that having a batch norm layer could speed up the model inference time?

**Comment by HuJiawei24** [comment]
My main concern is that why full precision model has comparable speed/inference time as compared to the BNN model.

**Comment by HuJiawei24** [comment]
The goal of my experiment is to see the inference time difference between the binary model and full precision model.
The experiment setup is the following: I chose resnet20 architecture, I remove all the bn and fc layers. Only con layer is reserved in the network. I used larq to create resnet20 binary version and full precision version. Then I use LCE to convert the .h5 file into tflite file. These two tflite files are subsequently executed/profiled on raspberry pi. And I found that the inference time between the binary model and full precision model is very negligible.
What could be the possible reasons behind this result?

**Comment by jneeven** [comment]
> And I found that the inference time between the binary model and full precision model is very negligible.
> What could be the possible reasons behind this result?

Hi! I had a look at the tflite files you linked, and benchmarked them on a Raspberry Pi 4B (the same we use for the benchmarks in [this table](https://docs.larq.dev/compute-engine/#performance)).

Averaging the results over 50 runs, I get the following numbers:
```
Full precision: 17.62ms
Binary: 4.78ms
```

So the speedup is indeed roughly what you would expect. I have also ran these benchmarks on a Pixel 1, and the ratio between the inference times is similar there.
This therefore appears not to be a problem with the model(s), but with the setup, and I suspect this issue might be related to https://github.com/larq/compute-engine/issues/383. Could you try converting and running one of the tflite files from [our zoo](https://docs.larq.dev/compute-engine/#performance) (e.g. QuickNet) and check if the inference time is roughly correct? If not, let's continue the investigation under that [LCE issue](https://github.com/larq/compute-engine/issues/383) instead.

That being said, as @leonoverweel mentioned the architecture of a BNN usually differs somewhat from its full-precision counterpart, so I'm not sure how informative the inference comparison between these two models would be (your current binarized model is unlikely to be the model that you would ever use in practice).
However, two tips I can give you regarding the BNN you sent:
- It currently uses zero-padding, which is relatively slow. Setting `pad_values=1.0` on your [QuantConv2D](https://docs.larq.dev/larq/api/layers/#quantconv2d-class) layers should speed up inference a bit.
- In LCE, we support fusing BatchNorms and ReLUs into binarized convolutions in the following order:
`QuantConv2D -> ReLu -> BatchNorm`. If you swap the ReLU and the BatchNorm (or, in your case, remove the BatchNorm altogether), the ReLU can no longer be fused (you can see the ReLU being a separate operation if you visualize your BNN tflite file with [netron](https://github.com/lutzroeder/netron)) and will slow down inference. You can either remove the ReLU or add the BatchNorm to resolve this issue (and I suggest adding the BatchNorm, it should greatly improve accuracy).

**Comment by HuJiawei24** [comment]
The os I used on raspberry pi is raspbian which is 32-bit os. Might this be the reason why the tested inference time for BNN is similar to that of the full precision model?

**Comment by AdamHillier** [comment]
> The os I used on raspberry pi is raspbian which is 32-bit os. Might this be the reason why the tested inference time for BNN is similar to that of the full precision model?

Ah thank you for raising that, that's a very good point. Our optimised Arm kernels are implemented for 64-bit Arm, so when using a 32-bit OS LCE falls back to reference kernels that will be quite a bit slower. The 64-bit OS we use on our Raspberry Pi 4B boards is [Manjaro](https://manjaro.org/download/arm/raspberry-pi-4/arm8-raspberry-pi-4-kde-plasma/), though any other 64-bit OS should work as well.

**Comment by lgeiger** [comment]
@HuJiawei24 Did using a 64-bi OS as suggested in https://github.com/larq/larq/issues/496#issuecomment-635299659 solve the issue for you?

**Comment by lgeiger** [comment]
I marked this as a duplicate of https://github.com/larq/compute-engine/issues/388

@HuJiawei24 Feel free to reopen if you still run into this issue.
